src/vision/status_quo/barbara_wants_single_threaded_optimizations.md

```diff
@@ -0,0 +1,126 @@
+
+impl ClientBroker {
+    fn start_serving_clients(self) {
+        let mut ipc_handle = self.make_ipc_handle_for_new_clients();
+        ipc_handle.spawn_on_reply(move |reply| shared_state.clone().serve_client(reply));
+        LocalExecutor::new().run_singlethreaded(ipc_handle.listen());
+    }
+
+    fn make_ipc_handle_for_new_clients(&self) { /* ... */ }
+    async fn serve_client(self, reply: IpcReply) { /* accesses interior mutability... */ }
+}
+```
+
+In order to be compatible with her own crate, Barbara needs to wrap the shared state of her service behind multi-threaded reference counting and synchronization:
+
+```rust
+#[derive(Clone)]
+struct ClientBroker {
+    state: Arc<Mutex<ClientState>>,
+}
+
+impl ClientBroker { /* nothing changed */ }
+```
+
+This incurs some performance overhead when cloning the `Arc` and when accessing the `Mutex`. The former is cheap when uncontended on x86 but will have different performance characteristics on e.g. ARM platforms. The latter's overhead varies depending on the kind of `Mutex` used, e.g. an uncontended `parking_lot::Mutex` may only need a few atomic instructions to acquire it. Acquiring many platforms' `std::sync::Mutex` is much more expensive than a few atomics. This overhead is usually not very high, but it does pollute shared resources like cache lines and is multiplied by the number of single-threaded services which make use of such a library.
+
+## 🤔 Frequently Asked Questions
+
+### **What are the morals of the story?**
+
+In synchronous Rust, choosing the "`Send`ness" of a crate is basically a choice about the concurrency it can support. In asynchronous Rust, one can write highly concurrent programs that still execute using only a single thread, but it is difficult to achieve maximum performance with reusable code.
+
+Abstracting over a library's `Send`ness requires being generic over storage/synchronization types *and* APIs which need to accept user-defined types/tasks/callbacks.
+
+### **What are the sources for this story?**
+
+As of writing, the [Fuchsia](https://fuchsia.dev) operating system had [over 1,500 invocations][st-invocations] of `LocalExecutor::run_singlethreaded`. There were [less than 500 invocations][mt-invocations] of `SendExecutor::run`.[^fuchsia-methods] As of writing the author could not find any widely used support libraries which were not thread-safe.
+
+[st-invocations]: https://cs.opensource.google/search?q=file:rs%20run_singlethreaded&sq=&ss=fuchsia%2Ffuchsia
+[mt-invocations]: https://cs.opensource.google/search?q=file:rs%20%5C.run%5C(&ss=fuchsia%2Ffuchsia
+
+`actix-rt`'s [spawn function](https://docs.rs/actix-rt/1.1.1/actix_rt/fn.spawn.html) does not require `Send` for its futures, because each task is polled on the thread that spawned it. However it is very common when using `actix-rt` via `actix-web` to make use of async crates originally designed for `tokio`, whose [spawn function](https://docs.rs/tokio/1.6.1/tokio/fn.spawn.html) does require `Send`.
```

Comment:
Again, to drive the point home maybe finish with "...which results in applications paying for thread safety that they don't use"?